frontend/src/state/js-dispatcher.ts

```diff
@@ -0,0 +1,58 @@
+import { plainToInstance } from "class-transformer";
+import { JsMessageType, messageConstructorMap, JsMessage } from "../utilities/js-messages";
+import type { RustEditorInstance, WasmInstance } from "./wasm-loader";
+
+type JsMessageCallback<T extends JsMessage> = (responseData: T) => void;
+type JsMessageCallbackMap = {
+	// Don't know a better way of typing this since it can be any subclass of JsMessage
+	// The functions interacting with this map are strongly typed though around JsMessage
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	[response: string]: JsMessageCallback<any> | undefined;
+};
+
+type Constructs<T> = new (...args: never[]) => T;
+
+type JSMessageFactory = (data: unknown, wasm: WasmInstance, instance: RustEditorInstance) => JsMessage;
+
+type MessageMaker = typeof JsMessage | JSMessageFactory;
+
+export class JsDispatcher {
+	private responseMap: JsMessageCallbackMap = {};
+
+	handleJsMessage(messageType: JsMessageType, responseData: Record<string, unknown>, wasm: WasmInstance, instance: RustEditorInstance) {
+		const messageConstructor = messageConstructorMap[messageType] as MessageMaker;
```

Comment:
This is not an unsafe cast. It is a cast from a specific union of exact types to a more generic type. It prevents typescript from complaining since it does not want to do the coercion automatically.

Reply:
`let a = b as X` is a type assertion. Like `any`, you're asking the compiler to trust you. It's unsafe.

`let a: X = b` is safe and checked.

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/consistent-type-assertions.md has an example (and is probably a good rule to add with all options set to `'never'`):

> `const x = {} as { foo: number }` will succeed

Reply:
@mfish33 I'm still not understanding your description of this as "not an unsafe cast". To the extent I understand my typescript, this is an unsafe cast. (See cauthmann's description above also.) If the TypeScript compiler can't figure out the types on its own, that means something is wrong with the types here. Do we need to specify the types of `export const messageConstructors = {` in `js-messages.ts`?